src/Storages/MergeTree/MergeTreePartsMover.cpp

```diff
@@ -272,7 +278,14 @@ MergeTreePartsMover::TemporaryClonedPart MergeTreePartsMover::clonePart(const Me
     cloned_part.part = std::move(builder).withPartFormatFromDisk().build();
     LOG_TRACE(log, "Part {} was cloned to {}", part->name, cloned_part.part->getDataPartStorage().getFullPath());
 
-    cloned_part.part->is_temp = data->allowRemoveStaleMovingParts();
+    cloned_part.part->is_temp = false;
+    if (data->allowRemoveStaleMovingParts())
+    {
+        cloned_part.part->is_temp = true;
+        /// Setting it in case connection to zookeeper is lost while moving
+        /// Otherwise part might be stuck in the moving directory due to the KEEPER_EXCEPTION in part's destructor
+        cloned_part.part->remove_tmp_policy = IMergeTreeDataPart::BlobsRemovalPolicyForTemporaryParts::REMOVE_BLOBS;
```

Comment:
Why we are sure that blobs is not used by cloned part in case of zero-copy?

Reply:
Didn't notice the case then zero copy part is fetched from other replica. In that case PRESERVE_BLOBS policy can be set because part is locking shared data only after it is [not temporary anymore](https://github.com/ClickHouse/ClickHouse/blob/2fce90ab76b851198b39041c9e7c4b03ad601ff7/src/Storages/MergeTree/MergeTreePartsMover.cpp#L306)